File: 1114 Print in Order.java
class Foo {
    
    boolean firstDone, secondDone; 
    Lock lock;
    
    public Foo() {
        firstDone = secondDone = false; 
        lock = new ReentrantLock(); 
    }

    public void first(Runnable printFirst) throws InterruptedException {
        // printFirst.run() outputs "first". Do not change or remove this line.
        synchronized(lock) {
            printFirst.run();
            firstDone = true; 
            lock.notifyAll(); 
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        // printSecond.run() outputs "second". Do not change or remove this line.
        synchronized(lock) {
            while(!firstDone) {
                lock.wait(); 
            } 
            printSecond.run();
            secondDone = true; 
            lock.notifyAll(); 
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        // printThird.run() outputs "third". Do not change or remove this line.
        synchronized(lock) {
            while(!firstDone || !secondDone) {
                lock.wait(); 
            } 
            printThird.run();
        }
    }
}

File: 1114 Print in Order.c
typedef struct {
    // User defined data may be declared here.
    volatile bool a;
    volatile bool b;
} Foo;

Foo* fooCreate() {
    Foo* obj = (Foo*) malloc(sizeof(Foo));
    obj->a = obj->b = false;
    // Initialize user defined data here.
    
    return obj;
}

void first(Foo* obj) {
    
    // printFirst() outputs "first". Do not change or remove this line.
    printFirst();
    obj -> a = true;
}

void second(Foo* obj) {
   while(1)
   {
       if(obj->a == true)
       {
           break;
       }
   }
    // printSecond() outputs "second". Do not change or remove this line.
    printSecond();
    obj->b = true;
}

void third(Foo* obj) {
 while(1)
   {
       if(obj->b == true)
       {
           break;
       }
   }    // printThird() outputs "third". Do not change or remove this line.
    printThird();
}

void fooFree(Foo* obj) {
    // User defined data may be cleaned up here.
    
    free(obj);
}
